modules/util/source-formatter/src/main/java/com/liferay/source/formatter/checks/JavaUpgradeVersionCheck.java

```diff
@@ -0,0 +1,516 @@
+
+		if (x == -1) {
+			return;
+		}
+
+		List<String> parameterList = JavaSourceUtil.getParameterList(
+			content.substring(x));
+
+		try {
+			Version toSchemaVersion = new Version(
+				StringUtil.removeChar(parameterList.get(1), CharPool.QUOTE));
+
+			if (isExcludedPath(
+					_JAVA_UPGRADE_PROCESS_EXCLUDES, absolutePath,
+					toSchemaVersion.toString())) {
+
+				return;
+			}
+
+			Version fromSchemaVersion = new Version(
+				StringUtil.removeChar(parameterList.get(0), CharPool.QUOTE));
+
+			String expectedIncrementType = _getExpectedIncrementType(
+				absolutePath, parameterList.subList(2, parameterList.size()),
+				imports, upgradePackageName);
+
+			Version expectedSchemaVersion = _getExpectedSchemaVersion(
+				fromSchemaVersion, expectedIncrementType);
+
+			if (expectedSchemaVersion.compareTo(toSchemaVersion) > 0) {
+				int lineNumber =
+					javaTerm.getLineNumber() + getLineNumber(content, x) - 1;
+
+				addMessage(
+					fileName, "Expected new version: " + expectedSchemaVersion,
+					lineNumber);
+			}
+		}
+		catch (IllegalArgumentException iae) {
+		}
+	}
+
+	private String _getColumnType(
+		String sql, String tableName, String columnName) {
+
+		int x = sql.indexOf("create table " + tableName);
+
+		if (x == -1) {
+			return null;
+		}
+
+		x = sql.indexOf(StringBundler.concat("\t", columnName, " "), x + 1);
+
+		if (x == -1) {
+			return null;
+		}
+
+		x = sql.indexOf(StringPool.SPACE, x + 1);
+
+		int y = x;
+
+		while (true) {
+			y = sql.indexOf(StringPool.SPACE, y + 1);
+
+			if (y == -1) {
+				return null;
+			}
+
+			String columnType = StringUtil.trim(sql.substring(x, y));
+
+			if (getLevel(columnType) == 0) {
+				return columnType;
+			}
+		}
+	}
+
+	private String _getExpectedIncrementType(
+			String absolutePath, List<String> upgradeSteps,
+			List<String> imports, String upgradePackageName)
+		throws Exception {
+
+		String incrementType = _INCREMENT_TYPE_MICRO;
+
+		for (String upgradeStep : upgradeSteps) {
+			if (upgradeStep.contains("{\n")) {
+				incrementType = _adjustIncrementTypeForJava(
+					absolutePath, upgradeStep, null, upgradePackageName,
+					incrementType);
+
+				if (incrementType.equals(_INCREMENT_TYPE_MAJOR)) {
+					return incrementType;
+				}
+
+				continue;
+			}
+
+			Matcher matcher = _classNamePattern.matcher(upgradeStep);
+
+			if (!matcher.find()) {
+				continue;
+			}
+
+			String className = StringUtil.removeChars(
+				matcher.group(1), CharPool.NEW_LINE, CharPool.SPACE,
+				CharPool.TAB);
+
+			if (!className.contains(StringPool.PERIOD)) {
+				for (String importName : imports) {
+					if (importName.endsWith(StringPool.PERIOD + className)) {
+						className = importName;
+
+						break;
+					}
+				}
+			}
+
+			if (!className.contains(StringPool.PERIOD)) {
+				className = StringBundler.concat(
+					upgradePackageName, StringPool.PERIOD, className);
+			}
+
+			if (!className.startsWith(upgradePackageName)) {
```

Comment:
Hi @hhuijser,

Why do we need to verify that the upgradeClass is in the upgrade package of the module?

Reply:
We do not.
For a first draft I did this for simplicity (easier to find the class). Removing this would make the check better.